File: Raytracer/Master.c
/**
 *  CPE 2010
 *  -------------------
 *  Program
 *
 *  Last Modified:
 *  @author Nick Feeney
 */
#include "Master.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

typedef char bool;
#define true 1
#define false 0

float planeHitTest( Plane plane,  Ray ray )
{
   vec3 direction = unit(ray.dir);
   vec3 position;
   vec3 normal = unit(plane.normal);
   direction.x = -direction.x;
   direction.y = -direction.y;
   direction.z = -direction.z;
   position.x = ray.pos.x;
   position.y = ray.pos.y;
   position.z = ray.pos.z;

   float vd = dot(normal, direction);
   if((distance < 0 && vd > -0.0001) || (distance > 0 && vd < 0.0001))
      return -1;
   float v0 = dot(newDirection(plane.point, position), plane.normal);
   float t = v0/vd;
   //make sure its pointing right directions
   if( t < 0 )
      t = -t;
   if( t < 0.001)
      return -1;
   return t;

}
Intersection planeIntersection(  Plane plane,  Ray ray, float t )
{
   Intersection ret;
   ret.hit = true;
   ret.viewVector.x = -ray.dir.x;
   ret.viewVector.y = -ray.dir.y;
   ret.viewVector.z = -ray.dir.z;

   ret.hitMark.x = ray.pos.x + ray.dir.x*t;
   ret.hitMark.y = ray.pos.y + ray.dir.y*t;
   ret.hitMark.z = ray.pos.z + ray.dir.z*t;
   ret.normal = plane.normal;
   ret.colorInfo = plane.info.colorInfo;
   return ret;
}

/**
 *  CPE 2010
 *  -------------------
 *  Program
 *
 *  Last Modified:
 *  @author Nick Feeney
 */

float sphereHitTest(  Sphere sphere,  Ray ray )
{
   vec3 direction = unit( ray.dir );
   float xc = sphere.pos.x;
   float yc = sphere.pos.y;
   float zc = sphere.pos.z;
   float x0 = ray.pos.x;
   float y0 = ray.pos.y;
   float z0 = ray.pos.z;
   float xd = direction.x;
   float yd = direction.y;
   float zd = direction.z;

   float A = xd*xd + yd*yd + zd*zd;
   float B = 2*(xd *(x0-xc) + yd*(y0-yc) + zd*(z0-zc));
   float C = (x0-xc)*(x0-xc) + (y0-yc)*(y0-yc) + (z0-zc)*(z0-zc) - sphere.radius*sphere.radius;
   float disc = B*B -4*A*C;
   if(disc < .0001)
      return -1;

   float t0 = (-B - sqrt(disc))/2;
   if ( t0 < 0.001)
   {
      t0 = (-B + sqrt(disc))/2;
   }
   if( t0 <= 0.001 )// && t0 >= -.00001 )
      return -1;
   return t0;
}
Intersection sphereIntersection(  Sphere sphere,  Ray ray, float t0 )
{
   vec3 direction = unit( ray.dir );
   Intersection ret;
   float xc = sphere.pos.x;
   float yc = sphere.pos.y;
   float zc = sphere.pos.z;
   float x0 = ray.pos.x;
   float y0 = ray.pos.y;
   float z0 = ray.pos.z;
   float xd = direction.x;
   float yd = direction.y;
   float zd = direction.z;
   ret.viewVector.x = -direction.x;
   ret.viewVector.y = -direction.y;
   ret.viewVector.z = -direction.z;
   ret.hitMark.x = x0 + xd*t0;
   ret.hitMark.y = y0 + yd*t0;
   ret.hitMark.z = z0 + zd*t0;

   vec3 objHit;
   objHit.x = ray.pos.x + direction.x* t0;
   objHit.y = ray.pos.y + direction.y* t0;
   objHit.z = ray.pos.z + direction.z* t0;

   ret.normal.x = (objHit.x - xc)/sphere.radius;
   ret.normal.y = (objHit.y - yc)/sphere.radius;
   ret.normal.z = (objHit.z - zc)/sphere.radius;
   ret.normal = unit(ret.normal);
   ret.colorInfo = sphere.info.colorInfo;

   ret.hit = true;
   return ret;
}


/**
 *  CPE 2010
 *  -------------------
 *  Program
 *
 *  Last Modified:
 *  @author Nick Feeney
 */

float triangleHitTest(  Triangle triangle,  Ray ray )
{
   vec3 direction = unit( ray.dir );
   vec3 position = ray.pos;
   direction.x = direction.x;
   direction.y = direction.y;
   direction.z = direction.z;

   float aa = triangle.a.x - triangle.b.x;
   float bb = triangle.a.y - triangle.b.y;
   float cc = triangle.a.z - triangle.b.z;
   float d = triangle.a.x - triangle.c.x;
   float e = triangle.a.y - triangle.c.y;
   float f = triangle.a.z - triangle.c.z;
   float g = direction.x;
   float h = direction.y;
   float i = direction.z;
   float j = triangle.a.x - position.x;
   float k = triangle.a.y - position.y;
   float l = triangle.a.z - position.z;

   float t = -1;
   float beta = 0;
   float gamma = 0;

   float ei_m_hf = e*i - h*f;
   float gf_m_di = g*f - d*i;
   float dh_m_eg = d*h - e*g;
   float ak_m_jb = aa*k - j*bb;
   float jc_m_al = j*cc - aa*l;
   float bl_m_kc = bb*l - k*cc;
   float M = aa*ei_m_hf + bb*gf_m_di + cc*dh_m_eg;
   if( M  < 0.0001 && M > -0.0001 )
   {
      return -1;
   }
   t = -(f*ak_m_jb + e*jc_m_al + d* bl_m_kc)/M;

   if(t < 0.001)
   {
      return -1;
   }

   gamma = (i*ak_m_jb + h*jc_m_al + g* bl_m_kc)/M;
   if(gamma < 0 || gamma > 1)
   {
      return -1;
   }

   beta = (j*ei_m_hf + k*gf_m_di + l*dh_m_eg)/M;
   if(beta < 0 || beta > (1 - gamma))
   {
      return -1;
   }

   return t;
}
Intersection triangleIntersection(  Triangle triangle,  Ray ray, float t )
{
   Intersection ret;

   ret.hit = true;
   vec3 direction = unit( ray.dir );
   ret.viewVector.x = -direction.x;
   ret.viewVector.y = -direction.y;
   ret.viewVector.z = -direction.z;
   ret.hitMark.x = ray.pos.x + direction.x*t;
   ret.hitMark.y = ray.pos.y + direction.y*t;
   ret.hitMark.z = ray.pos.z + direction.z*t;
   ret.normal = triangle.normal;
   ret.colorInfo = triangle.info.colorInfo;
   return ret;
}

/**
 *  CPE 2010
 *  -------------------
 *  Program
 *
 *  Last Modified:
 *  @author Nick Feeney
 */

#define PI 3.141592

ObjectInfo createObjectInfo()
{
   ObjectInfo obj;
   obj.colorInfo.pigment.r = 0;
   obj.colorInfo.pigment.g = 0;
   obj.colorInfo.pigment.b = 0;
   obj.colorInfo.pigment_f = 0;

   obj.colorInfo.finish_ambient = 0;
   obj.colorInfo.finish_diffuse = 0;
   obj.colorInfo.finish_specular = 0;
   obj.colorInfo.finish_roughness = 0;
   obj.colorInfo.finish_reflection = 0;
   obj.colorInfo.finish_refraction = 0;
   obj.colorInfo.finish_ior = 0;
   return obj;
}
/**
 *  CPE 2010
 *  -------------------
 *  Program
 *
 *  Last Modified:
 *  @author Nick Feeney
 */

#define PI 3.141592

int createInitRays( Ray **rays, int width, int height, Camera cam )
{
   vec3 right = unit(cam.right);
   vec3 up = unit(cam.up);
   float rightUnitX = right.x;
   float rightUnitY = right.y;
   float rightUnitZ = right.z;
   float upUnitX = up.x;
   float upUnitY = up.y;
   float upUnitZ = up.z;
   vec3 uv = unit(newDirection(cam.lookat, cam.pos));

   *rays = (Ray *) malloc( sizeof(Ray) *height*width );
   for( int i = 0; i < height; i++)
   {
      for( int j = 0; j < width ; j ++ )
      {
         float u = cam.l + (cam.r-cam.l)*( (float)j)/(float)width;
         float v = cam.b + (cam.t-cam.b)*( (float)i)/(float)height;
         float w = -1;
         int c = i*width + j;

         (*rays)[c].pos = cam.pos;
         (*rays)[c].dir.x = u * rightUnitX + v * upUnitX + -w * uv.x;
         (*rays)[c].dir.y = u * rightUnitY + v * upUnitY + -w * uv.y;
         (*rays)[c].dir.z = u * rightUnitZ + v * upUnitZ + -w * uv.z;
         (*rays)[c].i = i;
         (*rays)[c].j = j;
      }
   }
   return width * height;
}
void castRays( Scene scene, Ray *rays, int numRays, int width, int height, Color **buffer )
{
   for (int i = 0; i < numRays; i++)
       buffer[rays[i].i][rays[i].j] = raytrace( scene, rays[i] );
}
Color raytrace( Scene scene, Ray ray )
{
   Color color;
   color.r = 0;
   color.b = 0;
   color.g = 0;

   Intersection best;
   best.hit = false;

   float bestT = 10000;
   float t;
   for( int j = 0; j < scene.numSpheres; j++ )
   {
      t = sphereHitTest( scene.spheres[j], ray );
      if( t > 0 )
      {
         if( !best.hit || t < bestT )
         {
            best = sphereIntersection( scene.spheres[j], ray, t );
            bestT = t;
         }
      }
   }
   for( int j = 0; j < scene.numTriangles; j++ )
   {
      t = triangleHitTest( scene.triangles[j], ray );
      if( t > 0 )
      {
         if( !best.hit || t < bestT )
         {
            best = triangleIntersection( scene.triangles[j], ray, t );
            bestT = t;
         }
      }
   }
   for( int j = 0; j < scene.numPlanes; j++ )
   {
      t = planeHitTest( scene.planes[j], ray );
      if( t > 0 )
      {
         if( !best.hit || t < bestT )
         {
            best = planeIntersection( scene.planes[j], ray, t );
            bestT = t;
         }
      }
   }
   if( best.hit )
   {
      color = plus( color, directIllumination( best, scene ) );
      //printf("color: %f, %f, %f\n", color.r, color.g, color.b);
   }
   return limitColor( color );
}
Color limitColor(  Color in )
{
   Color ret;
   if( in.r > 1.0 )
      ret.r = 1.0;
   else if( in.r < 0.0 )
      ret.r = 0;
   else
      ret.r = in.r;

   if( in.g > 1.0 )
      ret.g = 1.0;
   else if( in.g < 0.0 )
      ret.g = 0;
   else 
      ret.g = in.g;

   if( in.b > 1.0 )
      ret.b = 1.0;
   else if( in.b < 0.0 )
      ret.b = 0;
   else
      ret.b = in.b;

   return ret;
}
Color plus(  Color first,  Color other )
{
   Color ret;
   ret.r = first.r + other.r;
   ret.g = first.g + other.g;
   ret.b = first.b + other.b;
   return limitColor( ret );
}

Color directIllumination( Intersection inter, Scene scene )
{
   Color ret;
   ret.r = 0;
   ret.b = 0;
   ret.g = 0;

   for( int i = 0; i < scene.numPointLights; i++ )
   {
      PointLight temp = scene.pointLights[i];
      vec3 lvec = unit(newDirection(temp.pos, inter.hitMark ));

      float nlDot = dot(lvec, inter.normal );
      bool inShadow = false;
      float lightDistance = distance( temp.pos, inter.hitMark );

      //contruct possible hits for shadow ray using bvh
      for( int j = 0; j < scene.numSpheres; j++ )
      {
         Ray shadowRay;
         shadowRay.pos = inter.hitMark;
         shadowRay.dir = lvec;
         float t = sphereHitTest(scene.spheres[j], shadowRay );
         if( t > 0 )
         {
            Intersection info = sphereIntersection( scene.spheres[j], shadowRay, t );
            if(info.hit)
            {
               if( distance( info.hitMark, inter.hitMark ) < lightDistance )
               {
                  inShadow = true;
                  break;
               }
            }
         }
      }

      ObjectInfo shadowInter;
      if( !inShadow )
      {
         vec3 r;
         r.x = -lvec.x + 2 * nlDot * inter.normal.x;
         r.y = -lvec.y + 2 * nlDot * inter.normal.y;
         r.z = -lvec.z + 2 * nlDot * inter.normal.z;
         r = unit(r);
         float rvDot = dot(r, unit(inter.viewVector));
         if(nlDot < 0)
            nlDot = 0;
         if(rvDot < 0)
            rvDot = 0;
         float powRV = pow( rvDot, 1.0/inter.colorInfo.finish_roughness );

         ret.r  =ret.r + temp.color.r * powRV*inter.colorInfo.finish_specular;
         ret.g = ret.g + temp.color.g* powRV*inter.colorInfo.finish_specular;
         ret.b = ret.b + temp.color.b * powRV*inter.colorInfo.finish_specular;
         ret.r+= inter.colorInfo.pigment.r * temp.color.r * nlDot*inter.colorInfo.finish_diffuse;
         ret.g += inter.colorInfo.pigment.g * temp.color.g * nlDot*inter.colorInfo.finish_diffuse;
         ret.b+= inter.colorInfo.pigment.b * temp.color.b * nlDot*inter.colorInfo.finish_diffuse;
         ret = limitColor( ret );
      }
   }
   //1.5 not 1 to increase the directlight which will be balanced in Util/tga.cpp during gamma correction
   float mod = 1;//.2 - finish_reflection - pigment_f*finish_refraction;
   ret.r = ret.r * mod;
   ret.g = ret.g * mod;
   ret.b = ret.b * mod;
   return ret;
}
/**
 *  CPE 2010
 *  -------------------
 *  Program
 *
 *  Last Modified:
 *  @author Nick Feeney
 */

float mag( vec3 in)
{
   return sqrt(in.x*in.x + in.y*in.y + in.z*in.z);
}
float dot( vec3 one,  vec3 two)
{
   return one.x*two.x + one.y*two.y + one.z*two.z;
}
vec3 cross( vec3 one, vec3 two)
{
   vec3 newVector;
   newVector.x = one.y*two.z - one.z*two.y;
   newVector.y = one.z*two.x - one.x*two.z;
   newVector.z = one.x*two.y - one.y*two.x;
   return newVector;
}
float theta( vec3 one,  vec3 two)
{
   return acosf( dot(one, two)/(mag(one) * mag(two))) * 180.0 / 3.14159;
}
float distance( vec3 one,  vec3 two )
{
   return sqrt((one.x-two.x)*(one.x-two.x) + (one.y-two.y)*(one.y-two.y) + (one.z-two.z)*(one.z-two.z));
}
vec3 newDirection( vec3 to,  vec3 from )
{
   vec3 newVec;
   newVec.x = to.x - from.x;
   newVec.y = to.y - from.y;
   newVec.z = to.z - from.z;
   return newVec;
}
vec3 unit( vec3 in)
{
   float temp;
   vec3 newVector;
   newVector.x = 0;
   newVector.y = 0;
   newVector.z = 0;
   temp = mag(in);

   if(temp > 0)
   {
      newVector.x = in.x/temp;
      newVector.y = in.y/temp;
      newVector.z = in.z/temp;
   }
   return newVector;
}

File: jsRayTrace/Master.js
var DIM = 400;

function vec3(x, y, z) {
	this.x = x;
	this.y = y;
	this.z = z;
}

function Camera(pos, up, right, lookAt, l, r, b, t) {
	this.pos = pos;
	this.up = up;
	this.right = right;
	this.lookAt = lookAt;
	this.l = l;
	this.r = r;
	this.b = b;
	this.t = t;
}

function Color(r, g, b, a) {
	this.r = r;
	this.g = g;
	this.b = b;
	this.a = a;
}

function PointLight(color, pos) {
	this.color = color; // Color
	this.pos = pos; // vec3
}

function ColorInfo(
		finish_ambient,
		finish_diffuse,
		finish_specular,
		finish_roughness,
		finish_reflection,
		finish_refraction,
		finish_ior,
		pigment_f,
		pigment) {
	this.finish_ambient = finish_ambient;
	this.finish_diffuse = finish_diffuse;
	this.finish_specular = finish_specular;
	this.finish_roughness = finish_roughness;
	this.finish_reflection = finish_reflection;
	this.finish_refraction = finish_refraction;
	this.finish_ior = finish_ior;
	this.pigment_f = pigment_f;
	this.pigment = pigment; // Color
}

function ObjectInfo(colorInfo) {
	this.colorInfo = colorInfo; // ColorInfo
}

function Plane(distancee, normal, point, info) {
	this.distancee = distancee;
	this.normal = normal; // vec3
	this.point = point; // vec3
	this.info = info; // ObjectInfo
}

function Sphere(radius, pos, info) {
	this.radius = radius;
	this.pos = pos; // vec3
	this.info = info; // ObjectInfo
}

function Triangle(distancee, a, b, c, normal, info) {
	this.distancee = distancee;
	this.a = a; // vec3
	this.b = b; // vec3
	this.c = c; // vec3
	this.normal = normal; // vec3
	this.info = info; // ObjectInfo
}

function Intersection(hitMark, normal, viewVector, hit, colorInfo) {
	this.hitMark = hitMark; // vec3
	this.normal = normal; // vec3
	this.viewVector = viewVector; // vec3
	this.hit = hit; // var
	this.colorInfo = colorInfo; // ColorInfo
}

function Ray(pos, dir, i, j) {
	this.pos = pos; // vec3
	this.dir = dir; // vec3
	this.i = i; // int
	this.j = j; // int
}

function Scene(spheres, planes, triangles, pointLights, camera) {
	this.spheres = spheres; // Sphere[]
	this.planes = planes; // Plane[]
	this.triangles = triangles; // Triangle[]
	this.pointLights = pointLights; // PointLight[]
	this.camera = camera; // Camera
}

  

function mag(inVec) {
	return Math.sqrt(inVec.x * inVec.x + inVec.y * inVec.y + inVec.z * inVec.z);
}
function dott(one, two) {
	return one.x * two.x + one.y * two.y + one.z * two.z;
}
function crosss(one, two) {
	return new vec3(
			one.y * two.z - one.z * two.y,
			one.z * two.x - one.x * two.z,
			one.x * two.y - one.y * two.x);
}
function theta(one, two) {
	return Math.acos(dott(one, two) / (mag(one) * mag(two))) * 180.0 / 3.14159;
}
function distancee(one, two) {
	return Math.sqrt((one.x-two.x)*(one.x-two.x) + (one.y-two.y)*(one.y-two.y) + (one.z-two.z)*(one.z-two.z));
}
function newDirection(to, from) {
	return new vec3(to.x - from.x, to.y - from.y, to.z - from.z);
}
function unit(inVec) {
	var temp = mag(inVec);
	if (temp > 0)
		return new vec3(inVec.x / temp, inVec.y / temp, inVec.z / temp);
	return new vec3(0, 0, 0);
}

function planeHitTest(plane, ray) {
	var direction = unit(ray.dir);
	direction = new vec3(-direction.x, -direction.y, -direction.z);
	var position = new vec3(ray.pos.x, ray.pos.y, ray.pos.z);
	var normal = unit(plane.normal);
	
	var vd = dott(normal, direction);
	if ((plane.distancee < 0 && vd > -0.0001) || (plane.distancee > 0 && vd < 0.0001))
		return -1;
	var v0 = dott(newDirection(plane.point, position), plane.normal);
	var t = v0 / vd;
	if (t < 0)
		t = -t;
	if (t < 0.001)
		return -1;
	return t;
}

function planeIntersection(plane, ray, t) {
	return new Intersection(
			new vec3(
					ray.pos.x + ray.dir.x * t,
					ray.pos.y + ray.dir.y * t,
					ray.pos.z + ray.dir.z * t),
			plane.normal,
			new vec3(-ray.dir.x, -ray.dir.y, -ray.dir.z),
			1,
			plane.info.colorInfo);
}

function sphereHitTest(sphere, ray) {
	var direction = unit(ray.dir);
	var xc = sphere.pos.x;
	var yc = sphere.pos.y;
	var zc = sphere.pos.z;
	var x0 = ray.pos.x;
	var y0 = ray.pos.y;
	var z0 = ray.pos.z;
	var xd = direction.x;
	var yd = direction.y;
	var zd = direction.z;
	
	var A = xd * xd + yd * yd + zd * zd;
	var B = 2 * (xd * (x0 - xc) + yd * (y0 - yc) + zd * (z0 - zc));
	var C = (x0-xc)*(x0-xc) + (y0-yc)*(y0-yc) + (z0-zc)*(z0-zc) - sphere.radius*sphere.radius;
	var disc = B * B - 4 * A * C;
	if (disc < .0001)
		return -1;
		
	var t0 = (-B - Math.sqrt(disc)) / 2;
	if (t0 < 0.001)
		t0 = (-B + Math.sqrt(disc)) / 2;
	if (t0 <= 0.001)
		return -1;
	return t0;
}

function sphereIntersection(sphere, ray, t0) {
	var direction = unit(ray.dir);
	var xc = sphere.pos.x;
	var yc = sphere.pos.y;
	var zc = sphere.pos.z;
	var x0 = ray.pos.x;
	var y0 = ray.pos.y;
	var z0 = ray.pos.z;
	var xd = direction.x;
	var yd = direction.y;
	var zd = direction.z;
	
   var objHit = new vec3();
   objHit.x = ray.pos.x + direction.x* t0;
   objHit.y = ray.pos.y + direction.y* t0;
   objHit.z = ray.pos.z + direction.z* t0;

	return new Intersection(
			new vec3(
					x0 + xd * t0,
					y0 + yd * t0,
					z0 + zd * t0),
			unit(new vec3(
					(objHit.x - xc) / sphere.radius,
					(objHit.y - yc) / sphere.radius,
					(objHit.z - zc) / sphere.radius)),
			new vec3(-direction.x, -direction.y, -direction.z),
			1,
			sphere.info.colorInfo);
}

function triangleHitTest(triangle, ray ) {
	var direction = new vec3(direction.x, direction.y, direction.z);
	var position = ray.pos;
	
	var aa = triangle.a.x - triangle.b.x;
	var bb = triangle.a.y - triangle.b.y;
	var cc = triangle.a.z - triangle.b.z;
	var d = triangle.a.x - triangle.c.x;
	var e = triangle.a.y - triangle.c.y;
	var f = triangle.a.z - triangle.c.z;
	var g = direction.x;
	var h = direction.y;
	var i = direction.z;
	var j = triangle.a.x - position.x;
	var k = triangle.a.y - position.y;
	var l = triangle.a.z - position.z;
	
	var t = -1;
	var beta = 0;
	var gamma = 0;
	
	var ei_m_hf = e*i - h*f;
	var gf_m_di = g*f - d*i;
	var dh_m_eg = d*h - e*g;
	var ak_m_jb = aa*k - j*bb;
	var jc_m_al = j*cc - aa*l;
	var bl_m_kc = bb*l - k*cc;
	var M = aa*ei_m_hf + bb*gf_m_di + cc*dh_m_eg;
	if( M  < 0.0001 && M > -0.0001 )
	{
		return -1;
	}
	t = -(f*ak_m_jb + e*jc_m_al + d* bl_m_kc)/M;
	
	if(t < 0.001)
	{
		return -1;
	}
	
	gamma = (i*ak_m_jb + h*jc_m_al + g* bl_m_kc)/M;
	if(gamma < 0 || gamma > 1)
	{
		return -1;
	}
	
	beta = (j*ei_m_hf + k*gf_m_di + l*dh_m_eg)/M;
	if(beta < 0 || beta > (1 - gamma))
	{
		return -1;
	}
	
	return t;
}

function triangleIntersection(triangle, ray, t) {
	return new Intersection(
			new vec3(
					ray.pos.x + direction.x * t,
					ray.pos.y + direction.y * t,
					ray.pos.z + direction.z * t),
			triangle.normal,
			new vec3(-direction.x, -direction.y, -direction.z),
			1,
			triangle.info.colorInfo);
}

var PI = 3.141592;

function createObjectInfo() {
	return new ObjectInfo(
			new ColorInfo(
					0, 0, 0, 0,
					0, 0,
					0,
					0,
					new Color(0, 0, 0)));
}

function createInitRays(rays, width, height, cam) {
   var right = unit(cam.right);
   var up = unit(cam.up);
   var rightUnitX = right.x;
   var rightUnitY = right.y;
   var rightUnitZ = right.z;
   var upUnitX = up.x;
   var upUnitY = up.y;
   var upUnitZ = up.z;
   var uv = unit(newDirection(cam.lookAt, cam.pos));

   //*rays = (Ray *) malloc( sizeof(Ray) *height*width );
   for( var i = 0; i < height; i++)
   {
      for( var j = 0; j < width ; j ++ )
      {
         var u = cam.l + (cam.r-cam.l)*j/width;
         var v = cam.b + (cam.t-cam.b)*i/height;
         var w = -1;
         var c = i*width + j;

					rays[c] = new Ray();
         rays[c].pos = cam.pos;
         rays[c].dir = new vec3();
         rays[c].dir.x = u * rightUnitX + v * upUnitX + -w * uv.x;
         rays[c].dir.y = u * rightUnitY + v * upUnitY + -w * uv.y;
         rays[c].dir.z = u * rightUnitZ + v * upUnitZ + -w * uv.z;
         rays[c].i = i;
         rays[c].j = j;
      }
   }
}
function clCastRays(rays, buffer)
{
	var scene = getScene();
	
	for (var i = 0; i < DIM * DIM; i++) {
		buffer[i] = raytrace(scene, rays[i]);
	}
}

function raytrace(scene, ray )
{
	var color = new Color(0, 0, 0, 1);
	
	var best = new Intersection();
	best.hit = 0;
   
   var bestT = 10000;
   var t = 0;
   
   for( var j = 0; j < scene.spheres.length; j++ )
   {
      t = sphereHitTest( scene.spheres[j], ray );
	  
      if( t > 0 )
      {
         if( !best.hit || t < bestT )
         {
            best = sphereIntersection( scene.spheres[j], ray, t );
            bestT = t;
         }
      }
	  
   }
   for( var j = 0; j < scene.triangles.length; j++ )
   {
      t = triangleHitTest( scene.triangles[j], ray );
      if( t > 0 )
      {
         if( !best.hit || t < bestT )
         {
            best = triangleIntersection( scene.triangles[j], ray, t );
            bestT = t;
         }
      }
   }
   for( var j = 0; j < scene.planes.length; j++ )
   {
      t = planeHitTest( scene.planes[j], ray );
      if( t > 0 )
      {
         if( !best.hit || t < bestT )
         {
            best = planeIntersection( scene.planes[j], ray, t );
            bestT = t;
         }
      }
   }
   if( best.hit )
   {
      color = plus( color, directIllumination( best, scene ) );
      //printf("color: %f, %f, %f\n", color.r, color.g, color.b);
   }
   return limitColor( color );
	
}
function limitColor( inc )
{
   var ret = new Color();
   
   if( inc.r > 1.0 )
      ret.r = 1.0;
   else if( inc.r < 0.0 )
      ret.r = 0;
   else
      ret.r = inc.r;

   if( inc.g > 1.0 )
      ret.g = 1.0;
   else if( inc.g < 0.0 )
      ret.g = 0;
   else 
      ret.g = inc.g;

   if( inc.b > 1.0 )
      ret.b = 1.0;
   else if( inc.b < 0.0 )
      ret.b = 0;
   else
      ret.b = inc.b;

   return ret;
}
function plus( first, other )
{
   var ret = new Color(
   		first.r + other.r,
			first.g + other.g,
			first.b + other.b);
   return limitColor( ret );
}

function directIllumination( inter, scene )
{
   ret = new Color(0, 0, 0, 1);

   for( var i = 0; i < scene.pointLights.length; i++ )
   {
      var temp = scene.pointLights[i];
      var lvec = unit(newDirection(temp.pos, inter.hitMark ));

      var nldott = dott(lvec, inter.normal );
      var inShadow = 0;
      var lightdistancee = distancee( temp.pos, inter.hitMark );

      //contruct possible hits for shadow ray using bvh
      for( var j = 0; j < scene.spheres.length; j++ )
      {
         var shadowRay = new Ray();
         shadowRay.pos = inter.hitMark;
         shadowRay.dir = lvec;
         var t = sphereHitTest(scene.spheres[j], shadowRay );
         if( t > 0 )
         {
            var info = sphereIntersection( scene.spheres[j], shadowRay, t );
            if(info.hit)
            {
               if( distancee( info.hitMark, inter.hitMark ) < lightdistancee )
               {
                  inShadow = 1;
                  break;
               }
            }
         }
      }

      var shadowInter = new ObjectInfo();
      if( !inShadow )
      {
         var r = new vec3();
         r.x = -lvec.x + 2 * nldott * inter.normal.x;
         r.y = -lvec.y + 2 * nldott * inter.normal.y;
         r.z = -lvec.z + 2 * nldott * inter.normal.z;
         r = unit(r);
         var rvdott = dott(r, unit(inter.viewVector));
         if(nldott < 0)
            nldott = 0;
         if(rvdott < 0)
            rvdott = 0;
         var powRV = Math.pow( rvdott, 1.0/inter.colorInfo.finish_roughness );

         ret.r  =ret.r + temp.color.r * powRV*inter.colorInfo.finish_specular;
         ret.g = ret.g + temp.color.g* powRV*inter.colorInfo.finish_specular;
         ret.b = ret.b + temp.color.b * powRV*inter.colorInfo.finish_specular;
         ret.r+= inter.colorInfo.pigment.r * temp.color.r * nldott*inter.colorInfo.finish_diffuse;
         ret.g += inter.colorInfo.pigment.g * temp.color.g * nldott*inter.colorInfo.finish_diffuse;
         ret.b+= inter.colorInfo.pigment.b * temp.color.b * nldott*inter.colorInfo.finish_diffuse;
         ret = limitColor( ret );
      }
   }
   //1.5 not 1 to increase the directlight which will be balanced in Util/tga.cpp during gamma correction
   var mod = 1;//.2 - finish_reflection - pigment_f*finish_refraction;
   ret.r = ret.r * mod;
   ret.g = ret.g * mod;
   ret.b = ret.b * mod;
   return ret;
}

function clRayTrace( rays, image) {
   scene = getScene();
   createInitRays( rays, DIM, DIM, scene.camera );
}
  
function getScene() {
	var right = new vec3(1.33, 0, 0);
	var up = new vec3(0, 1, 0);
	
	return new Scene(
			[
					new Sphere(
							2,
							new vec3(0, 0, 0),
							new ObjectInfo(
									new ColorInfo(
											.2, .4, 0, 0,
											0, 0,
											0,
											0,
											new Color(1.0, 0, 1.0, 1.0))))
			],
			[],
			[],
			[
					new PointLight(new Color(1.0, 1.0, 1.0, 1.0), new vec3(-100, 100, 100)),
			],
			new Camera(
					new vec3(0, 0, 14), // pos
					up, // up
					right, // right
					new vec3(0, 0, 0), // look at
					-mag(right) / 2,
					mag(right) / 2,
					-mag(up) / 2,
					mag(up) / 2));
}